### PR TITLE
main: add address encoding magic constants test

### PR DESCRIPTION
**Pull request by JFixby**
Ensures address encoding magics and `params.NetworkAddressPrefix` are consistent with each other. This test will light red when a new network is started with incorrect values.

Also there is a minor fix of *pointers for two other tests in this file.

### CODE REVIEW

**Comment by dnldd** [comment]
Good catch for the non-pointer `chaincfg.Params`, the failure conditions for the encoding tests are if the constants get modified or `CheckEncode` breaks right? If that's the case then I think you should look at adding tests to `github.com/decred/base58` instead since it's highly unlikely the defined constants will change.

**Comment by JFixby** [comment]
This test ensures address encoding magics and `params.NetworkAddressPrefix` are consistent with each other.

For example: the test will fail when someone changed `params.NetworkAddressPrefix` but forgot to update or updated incorrectly values in the `params.PubKeyAddrID`, `params.PubKeyHashAddrID`, `params.PKHEdwardsAddrID`,  `params.PKHSchnorrAddrID`, `params.ScriptHashAddrID` and `params.PrivateKeyID`.

Let's say a new network is started and the `params.NetworkAddressPrefix` is set to be `K`.
What would be the corresponding value for the `params.PubKeyAddrID`?
Is it `[0xfe, 0x07]`? Is the math correct?
The test will report when it is not.

File: networkparams_test.go [hunk 1]
```diff
@@ -70,13 +72,79 @@ func checkGenesisBlockRespectsNetworkPowLimit(
 	}
 }
 
+func checkPrefix(t *testing.T, prefix string, encoded, networkName string) {
```
Comment (davecgh) [inline networkparams_test.go]: Missing function comment.

File: networkparams_test.go [hunk 2]
```diff
@@ -49,7 +51,7 @@ func checkPowLimitsAreConsistent(t *testing.T, params chaincfg.Params) {
 //
 // This test ensures these blocks will respect the network PoW limit.
 func checkGenesisBlockRespectsNetworkPowLimit(
-	t *testing.T, params chaincfg.Params) {
+	t *testing.T, params *chaincfg.Params) {
```
Comment (davecgh) [inline networkparams_test.go]: Please put this on the same line as the function def so it is consistent with the rest.  Function definitions split on multiple lines causes some undesirable behavior with various tooling.

File: networkparams_test.go [hunk 3]
```diff
@@ -70,13 +72,79 @@ func checkGenesisBlockRespectsNetworkPowLimit(
...
+			"Address prefix mismatch for <%s>: expected <%s> received <%s>",
+			networkName,
+			prefix,
+			encoded,
+		)
+		t.FailNow()
+	}
+}
+
+func preFillArray(size int, value byte) []byte {
```
Comment (davecgh) [inline networkparams_test.go]: Missing function comment.

File: networkparams_test.go [hunk 4]
```diff
@@ -70,13 +72,79 @@ func checkGenesisBlockRespectsNetworkPowLimit(
...
+	return b
+}
+
+// checkInterval creates two corner cases defining interval
+// of all key values: [ xxxx000000000...0cccc , xxxx111111111...1cccc ],
+// where xxxx - is the encoding magic, and cccc is a checksum.
+// The interval is mapped to corresponding interval in base 58.
+// Then prefixes are checked for mismatch.
+func checkInterval(t *testing.T,
+	desiredPrefix string, keySize int, networkName string, magic [2]byte) {
```
Comment (davecgh) [inline networkparams_test.go]: Again, same line for func defs.

File: networkparams_test.go [hunk 5]
```diff
@@ -70,13 +72,79 @@ func checkGenesisBlockRespectsNetworkPowLimit(
...
+	// min and max possible keys
+	minKey := preFillArray(keySize, 0x00) // all zeroes
```
Comment (davecgh) [inline networkparams_test.go]: Please don't use postfix comments.  I know there are a couple of instances of it, but almost nothing in the codebase does and consistency is more important.

File: networkparams_test.go [hunk 6]
```diff
@@ -70,13 +72,79 @@ func checkGenesisBlockRespectsNetworkPowLimit(
 	}
 }
 
+func checkPrefix(t *testing.T, prefix string, encoded, networkName string) {
+	if strings.Index(encoded, prefix) != 0 {
+		t.Logf(
+			"Address prefix mismatch for <%s>: expected <%s> received <%s>",
```
Comment (davecgh) [inline networkparams_test.go]: Don't put single params per line like this.  It is not consistent with the rest of the code base.

File: networkparams_test.go (hunk 3 above)
Comment (davecgh) [inline networkparams_test.go]: This function is unncessary.  Use `bytes.Repeat` instead.

File: networkparams_test.go [hunk 7]
```diff
@@ -72,36 +72,27 @@ func checkGenesisBlockRespectsNetworkPowLimit(
...
-		t.Logf(
-			"Address prefix mismatch for <%s>: expected <%s> received <%s>",
-			networkName,
-			prefix,
-			encoded,
+// checkPrefix checks if targetString starts with the given prefix
+func checkPrefix(t *testing.T, prefix string, targetString, networkName string) {
+	if strings.Index(targetString, prefix) != 0 {
+		t.Logf("Address prefix mismatch for <%s>: expected <%s> received <%s>",
+			networkName, prefix, targetString,
```
Comment (davecgh) [inline networkparams_test.go]: Closing paren here.

Comment (JFixby) [reply networkparams_test.go]: Right.

**Comment by davecgh** [comment]
@JFixby I fixed it for you, but please note the code contribution guidelines for commits in the future.  Your commit message was not adhering to them in terms of max width.

**Comment by JFixby** [comment]
Thank you @davecgh.

I may also suggest making a pre-push checklist similar to [preflight checklist](https://en.wikipedia.org/wiki/Preflight_checklist).
Smth like:
- keep all code lines width within 80-90 chars limit, function contracts (defs) are ok to be longer
- no empty lines before the closing bracket of a function
- no empty lines after the opening bracket of a function
- don't use postfix comments
- ensure all functions have comments
- set commit message width below 80 chars
- check PR title width is below 80 chars
- set PR title package prefix

and so on.

**Comment by davecgh** [comment]
@JFixby There is a [Contribution Checklist](https://github.com/decred/dcrd/blob/master/docs/code_contribution_guidelines.md#Checklist) in the code contribution guidelines.  I believe all of this information is in that document.

If there is anything missing, I have no objections to PRs which add additional detail.

**Pull request merged**